spec/datadog/profiling/linux_tid_fallback_spec.rb

```diff
@@ -0,0 +1,104 @@
+      end
+
+      it 'always returns -1' do
+        expect(linux_tid_fallback_for(Thread.current)).to be(-1)
+      end
+    end
+
+    context 'when on Linux' do
+      before do
+        skip 'Test only runs on Linux' unless PlatformHelpers.linux?
+        check_for_process_vm_readv
+      end
+
+      context 'on Ruby >= 3.1' do
+        before { skip 'Behavior does not apply to current Ruby version' if RUBY_VERSION < '3.1.' }
+
+        # Thread#native_thread_id was added on 3.1
+        it 'returns the same as Thread#native_thread_id' do
+          all_threads = Thread.list
+
+          expect(all_threads.map(&:native_thread_id)).to eq(all_threads.map { |it| linux_tid_fallback_for(it) })
+        end
+      end
+
+      it 'returns the same as gettid' do
+        expect(linux_tid_fallback_for(Thread.current)).to be described_class::Testing._native_gettid
+      end
+    end
+  end
+
+  describe '.new_if_needed_and_working' do
+    subject(:new_if_needed_and_working) { described_class.new_if_needed_and_working }
+
+    context 'when not on Linux' do
+      before do
+        skip 'The fallback behavior only applies when not on Linux' if PlatformHelpers.linux?
+      end
+
+      it { is_expected.to be nil }
+    end
+
+    context 'when on Linux' do
+      before do
+        skip 'Test only runs on Linux' unless PlatformHelpers.linux?
+        check_for_process_vm_readv
+      end
+
+      context 'on Ruby >= 3.1' do
+        before { skip 'Behavior does not apply to current Ruby version' if RUBY_VERSION < '3.1.' }
+
+        it { is_expected.to be nil }
+      end
+
+      context 'on Ruby < 3.1' do
+        before { skip 'Behavior does not apply to current Ruby version' if RUBY_VERSION >= '3.1.' }
+
+        it { is_expected.to be_an_instance_of described_class }
+      end
+    end
+  end
+
+  # Why is this here? The LinuxTidFallback relies on the process_vm_readv Linux API, which needs special permissions
+  # and may be disabled. This is the case with CircleCI, where we usually run our tests.
+  #
+  # But it's dangerous to skip running specs in CI, since it may mean that we miss breakages. So, this check
+  # tries to be very precise: it only skips the specs IF indeed the API is not available + we're on CircleCI.
+  # (In case in the future our setup changes and the specs can run again.)
+  #
+  # It also breaks the specs with a clear error message when the API is not available; as otherwise the failures
+  # would be a bit cryptic.
+  #
+  # Finally, it adds a DD_PROFILING_SKIP_LINUX_TID_FALLBACK_TESTING that folks can use for local testing, in case
+  # their setup doesn't provide this API either.
+  def check_for_process_vm_readv
+    return if Datadog::Profiling::LinuxTidFallback::Testing._native_can_use_process_vm_readv?
+
+    if ENV['CIRCLECI'] == 'true'
+      skip "Skipping LinuxTidFallback specs because process_vm_readv doesn't work on CircleCI"
```

Comment:
Q: How hard would it be to support mocking `process_vm_readv`? You already seem to have native things dedicated to `::Testing`. If you accessed `process_vm_ready` through a function pointer, it sounds like you could have a `::Testing::mockProcessVMReadv` method that could replace that pointer during a CIRCLECI rspec run?  Might be better than skipping altogether?

Reply:
It would definitely be doable -- we already have the `struct linux_tid_fallback_state` that is stored inside instances of the `LinuxTidFallback` object, so we could store the right function pointer to use there. We could then introduce an initialize method for that class that had a flag to allow mocking, or something like that.

What makes me hesitate a bit is that's adding more complexity to the "production" code path, just for testing, whereas right now the extra stuff for testing is mostly off to the side.

Also, mocking would only catch regressions with not calling `LinuxTidFallback`, which... yeah I'm not sure how often it would happen.

So I'm a bit undecided on this one; do you think it's worth the effort for the extra bit of coverage?

Reply:
No strong opinion but I think mocking would also get you some CI coverage of the main behaviour on `:thread_id/uses the real native thread id` from `thread_contex_spec.rb` (unless that one works in CI and I'm just dumb)?

Without the mock, it seems like CI would be blind to deliberate or accidental changes to all of this functionality? Ofc the best approach would be to ensure some part of CI could actually meet the conditions to exercise this and get a full functional test  😅 

> that's adding more complexity to the "production" code path, just for testing, whereas right now the extra stuff for testing is mostly off to the side.

I understand the worry but OTOH, it sounds like the extra complexity would just be an extra pointer whose override would just be done under a testing run (similar to how you'd test a critical system using dependency injection), there'd be no extra testing-only loops or branches in production hotpath.

Reply:
I'm convinced, let me take a stab at it!